### PR TITLE
new module mso_backup

### PR DESCRIPTION
**Pull request by shrsr**
This module can:
Create, delete and query backups in local mode
 
Test file is included

### CODE REVIEW

File: plugins/modules/mso_backup.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    description = module.params.get('description')
+    location_type = module.params.get('location_type')
+    state = module.params.get('state')
+    backup = module.params.get('backup')
+
+    mso = MSOModule(module)
+
+    backup_id = None
+    path = 'backups'
+    count_backups = 0
```
Comment (lhercot) [inline plugins/modules/mso_backup.py:116]: remove count_backups

File: plugins/modules/mso_backup.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        if mso.existing:
+            data = mso.existing
+            mso.existing = []
+            for index_of_name in range(0, len(data)):
+                backup_id = data[index_of_name]['id']
+                if (str(backup) + '_' + backup_id == data[index_of_name]['metadata']['name']):
+                    path = 'backups/{id}'.format(id=backup_id)
+                    count_backups += 1
+                    mso.existing.append(data[index_of_name])
+            if count_backups < 1:
```
Comment (lhercot) [inline plugins/modules/mso_backup.py:128]: use len(mso.existing) here.

File: plugins/modules/mso_backup.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    mso.existing = mso.query_objs('backups')
+    if backup:
+        if mso.existing:
+            data = mso.existing
+            mso.existing = []
+            for index_of_name in range(0, len(data)):
+                backup_id = data[index_of_name]['id']
+                if (str(backup) + '_' + backup_id == data[index_of_name]['metadata']['name']):
+                    path = 'backups/{id}'.format(id=backup_id)
+                    count_backups += 1
```
Comment (lhercot) [inline plugins/modules/mso_backup.py:126]: Remove this line.

File: plugins/modules/mso_backup.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+    backup_id = None
+    path = 'backups'
+    count_backups = 0
+    mso.existing = mso.query_objs('backups')
+    if backup:
+        if mso.existing:
+            data = mso.existing
+            mso.existing = []
+            for index_of_name in range(0, len(data)):
```
Comment (lhercot) [inline plugins/modules/mso_backup.py:122]: This looks complicated. Isn't it easier to do:
```
for backup_info in data:
    backup_id = backup_info.get('id')
    if backup_info.get('metadata')['name'] == '{0}_{1}'.format(backup, backup_id):
        path = 'backups/{id}'.format(id=backup_id)
        mso.existing.append(backup_info)
```

File: plugins/modules/mso_backup.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                backup_id = data[index_of_name]['id']
+                if (str(backup) + '_' + backup_id == data[index_of_name]['metadata']['name']):
+                    path = 'backups/{id}'.format(id=backup_id)
+                    count_backups += 1
+                    mso.existing.append(data[index_of_name])
+            if count_backups < 1:
+                mso.existing = mso.get_obj(path, name=backup)
+
+    if state == 'query':
+        pass
```
Comment (lhercot) [inline plugins/modules/mso_backup.py:132]: why do pass and then exit_json?

File: plugins/modules/mso_backup.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        mso.exit_json()
+
+    if state == 'absent':
+        mso.previous = mso.existing
+        if mso.existing:
```
Comment (lhercot) [inline plugins/modules/mso_backup.py:137]: What happens if you state is absent but you have more than 1 backup returned? Your condition is probably not strong enough .

File: plugins/modules/mso_backup.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            if module.check_mode:
+                mso.existing = {}
+            else:
+                mso.existing = mso.request(path, method='DELETE')
+        mso.exit_json()
+
+    path = 'backups'
```
Comment (lhercot) [inline plugins/modules/mso_backup.py:144]: This looks like line 115 and 117 why redo this here? Why also not using get_obj before and just using query_objs?

File: plugins/modules/mso_user.py [hunk 8]
```diff
@@ -228,7 +228,6 @@ def main():
         payload = dict(
```
Comment (lhercot) [inline plugins/modules/mso_user.py:228]: Why do we have mso_user changes in this PR. It is in the other one right?
I am not reviewing those mso_user changes in this PR. Remove them and make sure they are in the other one.

File: tests/integration/targets/mso_backup/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,295 @@
+# Test code for the MSO modules
+# Copyright: (c) 2020, Lionel Hercot (@lhercot) <lhercot@cisco.com>
```
Comment (lhercot) [inline tests/integration/targets/mso_backup/tasks/main.yml:2]: Remove me and Dag from this file. I don't think we still use much of the shared code.

File: tests/integration/targets/mso_backup/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+    mso_info: &mso_info
+      host: '{{ mso_hostname }}'
+      username: '{{ mso_username }}'
+      password: '{{ mso_password }}'
+      validate_certs: '{{ mso_validate_certs | default(false) }}'
+      use_ssl: '{{ mso_use_ssl | default(true) }}'
+      use_proxy: '{{ mso_use_proxy | default(true) }}'
+      output_level: '{{ mso_output_level | default("info") }}'
+
+- name: Remove schemas
```
Comment (lhercot) [inline tests/integration/targets/mso_backup/tasks/main.yml:25]: Do you need any of this changes for testing the backup?
I think you can probably remove from Line25 to Line97.

File: tests/integration/targets/mso_backup/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+
+- name: Ensure Template 1 with AP1 exists
+  mso_schema_template_anp:
+    <<: *mso_info 
+    schema: '{{ mso_schema | default("ansible_test") }}'
+    template: Template 1
+    anp: AP1
+    state: present
+
+# Create Backup TO DO MSO Error?
```
Comment (lhercot) [inline tests/integration/targets/mso_backup/tasks/main.yml:98]: TO DO MSO Error?

File: tests/integration/targets/mso_backup/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+- name: Create Backup1 in check mode
+  mso_backup:
+    <<: *mso_info 
+    backup: Backup1
+    description: via Ansible
+    location_type: local
+    state: present
+  register: b_cm
```
Comment (lhercot) [inline tests/integration/targets/mso_backup/tasks/main.yml:106]: can you use longer variable names? I would use something like:cm_add_backup,  nm_add_backup, 
 cm_add_backup2
Make this change to all the test which don't have readable names.

File: tests/integration/targets/mso_backup/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+    schema: '{{ mso_schema | default("ansible_test") }}'
+    template: Template 1
+    anp: AP1
+    state: present
+
+# Create Backup TO DO MSO Error?
+- name: Create Backup1 in check mode
+  mso_backup:
+    <<: *mso_info 
+    backup: Backup1
```
Comment (lhercot) [inline tests/integration/targets/mso_backup/tasks/main.yml:102]: Could you prefix this with ansible_ like ansible_backup1?
This will avoid conflict in case of other people using same setup.
Apply this to the other cases too.

File: tests/integration/targets/mso_backup/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+  mso_backup:
+    <<: *mso_info 
+    state: query
+  ignore_errors: yes
+  register: b_qm_all
+
+- name: Verify b_qm_all
+  assert:
+    that:
+    - b_qm_all is not changed
```
Comment (lhercot) [inline tests/integration/targets/mso_backup/tasks/main.yml:196]: You should check the names of the backups you already know or at least that the number is there.

File: tests/integration/targets/mso_backup/tasks/main.yml [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+    state: present
+
+- name: Ensure Template 1 with AP1 exists
+  mso_schema_template_anp:
+    <<: *mso_info 
+    schema: '{{ mso_schema | default("ansible_test") }}'
+    template: Template 1
+    anp: AP1
+    state: present
+
```
Comment (lhercot) [inline tests/integration/targets/mso_backup/tasks/main.yml:97]: Shouldn't you have a step to clean previous backups with same name? Or it does not matter to your tests?

File: tests/integration/targets/mso_backup/tasks/main.yml [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+  ignore_errors: yes
+  
+- name: Remove Backup2
+  mso_backup:
+    <<: *mso_info 
+    backup: Backup2
+    state: absent
+  ignore_errors: yes
+
+# Add more data
```
Comment (lhercot) [inline tests/integration/targets/mso_backup/tasks/main.yml:267]: Why this test?

File: tests/integration/targets/mso_user/tasks/main.yml [hunk 17]
```diff
@@ -19,15 +19,13 @@
     use_ssl: '{{ mso_use_ssl | default(true) }}'
     use_proxy: '{{ mso_use_proxy | default(true) }}'
     output_level: '{{ mso_output_level | default("info") }}'
-    user: ansible_test
+    user: '{{ item }}'
```
Comment (lhercot) [inline tests/integration/targets/mso_user/tasks/main.yml:22]: Those changes should not be part of this PR.

File: plugins/modules/mso_backup.py (hunk 7 above)
Comment (shrsr) [reply plugins/modules/mso_backup.py:144]: I think mso.get_obj(path, name=backup) is empty because the name attribute in the response body is in metadata and not with id of the backup. Hence, I use query_objs..

"backups": [
    {
      "id": "20200718030540",
      "sizeKB": 790.19,
      "timestamp": "2020-07-18T03:05:40",
      "metadata": {
        "description": "",
        "name": "Backup100_20200718030540"
      }
    }

Comment (shrsr) [reply plugins/modules/mso_backup.py:144]: path = 'backups' is like a reset button in this case. We need this because we only use Post method, which requires the main path. If we had Put/Patch we could've avoided it

File: tests/integration/targets/mso_backup/tasks/main.yml (hunk 15 above)
Comment (shrsr) [reply tests/integration/targets/mso_backup/tasks/main.yml:97]: I did add removal steps at the bottom, but yeah I'll move it to the beginning. Will have to see how we can manipulate our next test file related to backups if any.

File: plugins/modules/mso_backup.py (hunk 7 above)
Comment (lhercot) [reply plugins/modules/mso_backup.py:144]: But then why do you use it here. It will clear out the mso.existing which you already have?
If path is always backup for the POST, just put it in the line 161 mso.request you don't need a variable single use.
